add new module: azure_rm_rediscachefirewallrule

yungezz opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
1. add  new module azure_rm_rediscachefirewallrule
2. change term "redis cache" to "Azure Cache for Redis" in existing redis modules.

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- New Module Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

yungezz committed:
add redis cache firewall rule

yungezz committed:
add test in alias

yungezz committed:
fix lint

zikalino commented on lib/ansible/modules/cloud/azure/azure_rm_rediscachefirewallrule.py:
```diff
@@ -0,0 +1,317 @@
+#!/usr/bin/python
+#
+# Copyright (c) 2018 Yunge Zhu, <yungez@microsoft.com>
```
2019 …. (@yungezz)

zikalino commented on lib/ansible/modules/cloud/azure/azure_rm_rediscachefirewallrule.py:
```diff
@@ -0,0 +1,317 @@
+version_added: "2.8"
+short_description: Manage Azure Cache for Redis Firewall rules.
+description:
+    - Create, update and delete instance of Azure Cache for Redis.
```
+ Firewall Rule
… maybe the name is getting to long?

zikalino commented on lib/ansible/modules/cloud/azure/azure_rm_rediscachefirewallrule.py:
```diff
@@ -0,0 +1,317 @@
+        )
+
+        self._client = None
+
```
I just wonder, maybe we should start adding "required_if"?
I think start_ip_address and end_ip_address are required when state is present?

yungezz replied:
when updating, only one is required, the other one is not. how to express it?

yungezz replied:
added description in doc to mention it's required when creating

zikalino commented on lib/ansible/modules/cloud/azure/azure_rm_rediscachefirewallrule.py:
```diff
@@ -0,0 +1,317 @@
+                self.to_do = Actions.Delete
+            else:
+                self.results['changed'] = False
+                self.log("Azure Cache for Redis {0} not exists.".format(self.name))
```
not exists -> doesn't exit

zikalino reviewed (approved):
looks good, just a few small comments

yungezz committed:
resolve comments

yungezz committed:
comment out test due to redis cache creation too long

yungezz commented:
ready_for_review

yungezz commented:
add manual test result for reference.
```
TASK [Create firewall rule (Check mode)] ******************************************************************************************************************************************************************
task path: /home/yungez/src/ansible-testapp/redis.yml:44
Monday 04 March 2019  10:59:50 +0800 (0:00:02.060)       0:00:03.576 **********
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: yungez
<127.0.0.1> EXEC /bin/sh -c 'echo ~yungez && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/yungez/.ansible/tmp/ansible-tmp-1551668390.1411326-142180409811593 `" && echo ansible-tmp-1551668390.1411326-142180409811593="` echo /home/yungez/.ansible/tmp/ansible-tmp-1551668390.1411326-142180409811593 `" ) && sleep 0'
Using module file /home/yungez/.ansible/roles/azure_preview_modules/library/azure_rm_rediscachefirewallrule.py
<127.0.0.1> PUT /home/yungez/.ansible/tmp/ansible-local-8449ywqf906r/tmp03rz3rc1 TO /home/yungez/.ansible/tmp/ansible-tmp-1551668390.1411326-142180409811593/AnsiballZ_azure_rm_rediscachefirewallrule.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/yungez/.ansible/tmp/ansible-tmp-1551668390.1411326-142180409811593/ /home/yungez/.ansible/tmp/ansible-tmp-1551668390.1411326-142180409811593/AnsiballZ_azure_rm_rediscachefirewallrule.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/home/yungez/src/ansible/venv/bin/python /home/yungez/.ansible/tmp/ansible-tmp-1551668390.1411326-142180409811593/AnsiballZ_azure_rm_rediscachefirewallrule.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/yungez/.ansible/tmp/ansible-tmp-1551668390.1411326-142180409811593/ > /dev/null 2>&1 && sleep 0'
 [WARNING]: Azure API profile latest does not define an entry for RedisManagementClient

changed: [localhost] => changed=true
  id: null
  invocation:
    module_args:
      ad_user: null
      adfs_authority_url: null
      api_profile: latest
      auth_source: null
      cache_name: testredis1aa
      cert_validation_mode: null
      client_id: null
      cloud_environment: AzureCloud
      end_ip_address: 192.168.1.4
      name: myrule1
      password: null
      profile: null
      resource_group: yungezrerdistest1
      secret: null
      start_ip_address: 192.168.1.1
      state: present
      subscription_id: null
      tenant: null

TASK [Assert check mode creation] *************************************************************************************************************************************************************************
task path: /home/yungez/src/ansible-testapp/redis.yml:54
Monday 04 March 2019  10:59:52 +0800 (0:00:02.815)       0:00:06.391 **********
ok: [localhost] => changed=false
  msg: All assertions passed

TASK [Create firewall rule] *******************************************************************************************************************************************************************************
task path: /home/yungez/src/ansible-testapp/redis.yml:59
Monday 04 March 2019  10:59:52 +0800 (0:00:00.057)       0:00:06.449 **********
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: yungez
<127.0.0.1> EXEC /bin/sh -c 'echo ~yungez && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/yungez/.ansible/tmp/ansible-tmp-1551668393.0141082-235328291934119 `" && echo ansible-tmp-1551668393.0141082-235328291934119="` echo /home/yungez/.ansible/tmp/ansible-tmp-1551668393.0141082-235328291934119 `" ) && sleep 0'
Using module file /home/yungez/.ansible/roles/azure_preview_modules/library/azure_rm_rediscachefirewallrule.py
<127.0.0.1> PUT /home/yungez/.ansible/tmp/ansible-local-8449ywqf906r/tmp3pptytq6 TO /home/yungez/.ansible/tmp/ansible-tmp-1551668393.0141082-235328291934119/AnsiballZ_azure_rm_rediscachefirewallrule.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/yungez/.ansible/tmp/ansible-tmp-1551668393.0141082-235328291934119/ /home/yungez/.ansible/tmp/ansible-tmp-1551668393.0141082-235328291934119/AnsiballZ_azure_rm_rediscachefirewallrule.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/home/yungez/src/ansible/venv/bin/python /home/yungez/.ansible/tmp/ansible-tmp-1551668393.0141082-235328291934119/AnsiballZ_azure_rm_rediscachefirewallrule.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/yungez/.ansible/tmp/ansible-tmp-1551668393.0141082-235328291934119/ > /dev/null 2>&1 && sleep 0'
changed: [localhost] => changed=true
  id: /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/yungezrerdistest1/providers/Microsoft.Cache/redis/testredis1aa/firewallRules/myrule1
  invocation:
    module_args:
      ad_user: null
      adfs_authority_url: null
      api_profile: latest
      auth_source: null
      cache_name: testredis1aa
      cert_validation_mode: null
      client_id: null
      cloud_environment: AzureCloud
      end_ip_address: 192.168.1.4
      name: myrule1
      password: null
      profile: null
      resource_group: yungezrerdistest1
      secret: null
      start_ip_address: 192.168.1.1
      state: present
      subscription_id: null
      tenant: null

TASK [Assert creation] ************************************************************************************************************************************************************************************
task path: /home/yungez/src/ansible-testapp/redis.yml:68
Monday 04 March 2019  10:59:58 +0800 (0:00:05.080)       0:00:11.529 **********
ok: [localhost] => changed=false
  msg: All assertions passed

TASK [Update firewall rule idempotence] *******************************************************************************************************************************************************************
task path: /home/yungez/src/ansible-testapp/redis.yml:74
Monday 04 March 2019  10:59:58 +0800 (0:00:00.056)       0:00:11.586 **********
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: yungez
<127.0.0.1> EXEC /bin/sh -c 'echo ~yungez && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/yungez/.ansible/tmp/ansible-tmp-1551668398.1490552-168668289871837 `" && echo ansible-tmp-1551668398.1490552-168668289871837="` echo /home/yungez/.ansible/tmp/ansible-tmp-1551668398.1490552-168668289871837 `" ) && sleep 0'
Using module file /home/yungez/.ansible/roles/azure_preview_modules/library/azure_rm_rediscachefirewallrule.py
<127.0.0.1> PUT /home/yungez/.ansible/tmp/ansible-local-8449ywqf906r/tmp45q08u30 TO /home/yungez/.ansible/tmp/ansible-tmp-1551668398.1490552-168668289871837/AnsiballZ_azure_rm_rediscachefirewallrule.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/yungez/.ansible/tmp/ansible-tmp-1551668398.1490552-168668289871837/ /home/yungez/.ansible/tmp/ansible-tmp-1551668398.1490552-168668289871837/AnsiballZ_azure_rm_rediscachefirewallrule.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/home/yungez/src/ansible/venv/bin/python /home/yungez/.ansible/tmp/ansible-tmp-1551668398.1490552-168668289871837/AnsiballZ_azure_rm_rediscachefirewallrule.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/yungez/.ansible/tmp/ansible-tmp-1551668398.1490552-168668289871837/ > /dev/null 2>&1 && sleep 0'
ok: [localhost] => changed=false
  id: /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/yungezrerdistest1/providers/Microsoft.Cache/Redis/testredis1aa/firewallRules/myrule1
  invocation:
    module_args:
      ad_user: null
      adfs_authority_url: null
      api_profile: latest
      auth_source: null
      cache_name: testredis1aa
      cert_validation_mode: null
      client_id: null
      cloud_environment: AzureCloud
      end_ip_address: 192.168.1.4
      name: myrule1
      password: null
      profile: null
      resource_group: yungezrerdistest1
      secret: null
      start_ip_address: 192.168.1.1
      state: present
      subscription_id: null
      tenant: null

TASK [Assert idempotence] *********************************************************************************************************************************************************************************
task path: /home/yungez/src/ansible-testapp/redis.yml:83
Monday 04 March 2019  11:00:00 +0800 (0:00:02.675)       0:00:14.261 **********
ok: [localhost] => changed=false
  msg: All assertions passed

TASK [Update firewall rule] *******************************************************************************************************************************************************************************
task path: /home/yungez/src/ansible-testapp/redis.yml:88
Monday 04 March 2019  11:00:00 +0800 (0:00:00.056)       0:00:14.318 **********
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: yungez
<127.0.0.1> EXEC /bin/sh -c 'echo ~yungez && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/yungez/.ansible/tmp/ansible-tmp-1551668400.8833194-261491128807110 `" && echo ansible-tmp-1551668400.8833194-261491128807110="` echo /home/yungez/.ansible/tmp/ansible-tmp-1551668400.8833194-261491128807110 `" ) && sleep 0'
Using module file /home/yungez/.ansible/roles/azure_preview_modules/library/azure_rm_rediscachefirewallrule.py
<127.0.0.1> PUT /home/yungez/.ansible/tmp/ansible-local-8449ywqf906r/tmpsd0erqwu TO /home/yungez/.ansible/tmp/ansible-tmp-1551668400.8833194-261491128807110/AnsiballZ_azure_rm_rediscachefirewallrule.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/yungez/.ansible/tmp/ansible-tmp-1551668400.8833194-261491128807110/ /home/yungez/.ansible/tmp/ansible-tmp-1551668400.8833194-261491128807110/AnsiballZ_azure_rm_rediscachefirewallrule.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/home/yungez/src/ansible/venv/bin/python /home/yungez/.ansible/tmp/ansible-tmp-1551668400.8833194-261491128807110/AnsiballZ_azure_rm_rediscachefirewallrule.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/yungez/.ansible/tmp/ansible-tmp-1551668400.8833194-261491128807110/ > /dev/null 2>&1 && sleep 0'
changed: [localhost] => changed=true
  id: /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/yungezrerdistest1/providers/Microsoft.Cache/redis/testredis1aa/firewallRules/myrule1
  invocation:
    module_args:
      ad_user: null
      adfs_authority_url: null
      api_profile: latest
      auth_source: null
      cache_name: testredis1aa
      cert_validation_mode: null
      client_id: null
      cloud_environment: AzureCloud
      end_ip_address: 192.168.1.5
      name: myrule1
      password: null
      profile: null
      resource_group: yungezrerdistest1
      secret: null
      start_ip_address: null
      state: present
      subscription_id: null
      tenant: null

TASK [Assert updating] ************************************************************************************************************************************************************************************
task path: /home/yungez/src/ansible-testapp/redis.yml:96
Monday 04 March 2019  11:00:06 +0800 (0:00:05.314)       0:00:19.633 **********
ok: [localhost] => changed=false
  msg: All assertions passed

TASK [Delete firewall rule] *******************************************************************************************************************************************************************************
task path: /home/yungez/src/ansible-testapp/redis.yml:101
Monday 04 March 2019  11:00:06 +0800 (0:00:00.056)       0:00:19.690 **********
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: yungez
<127.0.0.1> EXEC /bin/sh -c 'echo ~yungez && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/yungez/.ansible/tmp/ansible-tmp-1551668406.2560713-246717278807557 `" && echo ansible-tmp-1551668406.2560713-246717278807557="` echo /home/yungez/.ansible/tmp/ansible-tmp-1551668406.2560713-246717278807557 `" ) && sleep 0'
Using module file /home/yungez/.ansible/roles/azure_preview_modules/library/azure_rm_rediscachefirewallrule.py
<127.0.0.1> PUT /home/yungez/.ansible/tmp/ansible-local-8449ywqf906r/tmpko28r8l1 TO /home/yungez/.ansible/tmp/ansible-tmp-1551668406.2560713-246717278807557/AnsiballZ_azure_rm_rediscachefirewallrule.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/yungez/.ansible/tmp/ansible-tmp-1551668406.2560713-246717278807557/ /home/yungez/.ansible/tmp/ansible-tmp-1551668406.2560713-246717278807557/AnsiballZ_azure_rm_rediscachefirewallrule.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/home/yungez/src/ansible/venv/bin/python /home/yungez/.ansible/tmp/ansible-tmp-1551668406.2560713-246717278807557/AnsiballZ_azure_rm_rediscachefirewallrule.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/yungez/.ansible/tmp/ansible-tmp-1551668406.2560713-246717278807557/ > /dev/null 2>&1 && sleep 0'
changed: [localhost] => changed=true
  id: /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/yungezrerdistest1/providers/Microsoft.Cache/Redis/testredis1aa/firewallRules/myrule1
  invocation:
    module_args:
      ad_user: null
      adfs_authority_url: null
      api_profile: latest
      auth_source: null
      cache_name: testredis1aa
      cert_validation_mode: null
      client_id: null
      cloud_environment: AzureCloud
      end_ip_address: null
      name: myrule1
      password: null
      profile: null
      resource_group: yungezrerdistest1
      secret: null
      start_ip_address: null
      state: absent
      subscription_id: null
      tenant: null

TASK [Assert deletion] ************************************************************************************************************************************************************************************
task path: /home/yungez/src/ansible-testapp/redis.yml:109
Monday 04 March 2019  11:00:11 +0800 (0:00:04.836)       0:00:24.527 **********
ok: [localhost] => changed=false
  msg: All assertions passed

```

Fred-sun commented:
@yuwzho @nitzmahone @jborean93 review_need

nitzmahone reviewed (changes requested):
Module content looks fine, but we should probably do something a little different for the tests so they can be run manually without modification. Since there's one check-mode test that can run, I'm not sure what the best way to handle the disabled tests is... Ideally they wouldn't be commented out, just disabled conditionally for CI. @mattclay - do we have a sane way to pass a condition to ansible-test to manually enable a broader set of tests for a target?

mattclay commented:
@nitzmahone Probably the easiest option would be to use tags and pass the `--tags` option to `ansible-test` as needed to run the slow tests.

Fred-sun commented:
@mattclay Do you mean adding tags to the test? Thanks!

Fred-sun commented:
@yuwzho Please help review this PR and push for forward! Thanks!

nitzmahone reviewed (changes requested):
switch commented out tests to use a tag instead, then +1

yungezz commented:
HI @mattclay  could you pls share how to add tags to test? is it Azure resource tag? or some ansible tag?  thanks in advance.

yungezz committed:
add tags to test

yungezz committed:
fix typo

yungezz commented:
hi @mattclay  I added tag `long_run` in test. Is it ok? Or let me know you have some specific tag for it. Thanks.

nitzmahone reviewed (approved):
changes LGTM; tag name isn't really important so long as the person running the full tests manually knows it, and it's unlikely to conflict with something defined by the test infra

Fred-sun commented:
@yungezz Please help resolve the conflicting files when you're available! Thanks!

yungezz committed:
fix merge issue

mattclay commented on test/integration/targets/azure_rm_rediscache/tasks/main.yml:
```diff
@@ -190,9 +191,115 @@
+  retries: 30
+  delay: 60
+  tags:
+    - long_run
```
The `never` tag is also needed. Otherwise the task still runs by default.

The same applies for all the other tasks with `long_run`.

yungezz replied:
sure. will link pr here when ready.

yungezz replied:
Matt fixed it already. thanks Matt.

mattclay reviewed (commented):
@yungezz Sorry for the late review. Can you open a PR to fix the tags?
